staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go

```diff
@@ -136,12 +155,29 @@ func (az *Cloud) CreateRoute(ctx context.Context, clusterName string, nameHint s
 	if err := az.createRouteTableIfNotExists(clusterName, kubeRoute); err != nil {
 		return err
 	}
-	targetIP, _, err := az.getIPForMachine(kubeRoute.TargetNode)
-	if err != nil {
-		return err
-	}
+	if !utilfeature.DefaultFeatureGate.Enabled(IPv6DualStack) {
+		targetIP, _, err = az.getIPForMachine(kubeRoute.TargetNode)
+		if err != nil {
+			return err
+		}
+	} else {
+		// for dual stack we need to select
+		// a private ip that matches family of the cidr
+		klog.V(4).Infof("CreateRoute: create route instance=%q cidr=%q is in dual stack mode", kubeRoute.TargetNode, kubeRoute.DestinationCIDR)
+		bCIDRv6 := utilnet.IsIPv6CIDRString(string(kubeRoute.DestinationCIDR))
+		nodePrivateIPs, err := az.getPrivateIPsForMachine(kubeRoute.TargetNode)
+		if nil != err {
+			klog.V(3).Infof("CreateRoute: create route: failed(GetPrivateIPsByNodeName) instance=%q cidr=%q with error=%v", kubeRoute.TargetNode, kubeRoute.DestinationCIDR, err)
+			return err
+		}
 
-	routeName := mapNodeNameToRouteName(kubeRoute.TargetNode)
+		targetIP, err = findFirstIPByFamily(nodePrivateIPs, bCIDRv6)
```

Comment:
Do we need to create routes for both ipv4 and ipv6? both IP families should be routable, right?

Reply:
`route controller` takes care of this. It walks `Node.CIDRs` and performs `CreateRoute()` per each. The route creation has to components `cidr` and `target node not target IP, hence the cloud provider needs to match family to node IP.